api/metric/api.go

```diff
@@ -16,66 +16,97 @@ package metric
 
 import (
 	"context"
+	"time"
 
 	"go.opentelemetry.io/api/core"
 	"go.opentelemetry.io/api/registry"
 	"go.opentelemetry.io/api/unit"
 )
 
-type MetricType int
+// Kind categorizes different kinds of metric.
+type Kind int
 
+//go:generate stringer -type=Kind
 const (
-	Invalid    MetricType = iota
-	Gauge                 // Supports Set()
-	Cumulative            // Supports Inc()
+	Invalid        Kind = iota
+	CumulativeKind      // Supports Add()
+	GaugeKind           // Supports Set()
+	MeasureKind         // Supports Record()
 )
 
-type Meter interface {
-	// TODO more Metric types
-	GetFloat64Gauge(ctx context.Context, gauge *Float64GaugeHandle, labels ...core.KeyValue) Float64Gauge
+// Recorder is the implementation-level interface to Set/Add/Record individual metrics.
+type Recorder interface {
+	// RecordAt allows the SDK to observe a single metric event
+	RecordAt(ctx context.Context, t time.Time, value float64)
 }
 
-type Float64Gauge interface {
-	Set(ctx context.Context, value float64, labels ...core.KeyValue)
+// Meter is an interface to the metrics portion of the OpenTelemetry SDK.
+type Meter interface {
+	// RecorderFor returns a handle for observing single measurements.
+	RecorderFor(context.Context, Instrument, ...core.KeyValue) Recorder
+
+	// RecordBatch atomically records a batch of measurements.
+	RecordBatch(context.Context, ...Measurement)
 }
 
-type Handle struct {
+// Instrument represents a named metric with recommended local-aggregation keys.
+type Instrument struct {
+	// Variable defines the metric name, description, and unit.
 	Variable registry.Variable
 
-	Type MetricType
+	// Kind is the metric kind of this instrument.
+	Kind Kind
+
+	// Keys are recommended keys determined in the handles
+	// obtained for this metric.
 	Keys []core.Key
 }
 
-type Option func(*Handle, *[]registry.Option)
+// Handle contains a Recorder to support the implementation-defined
+// behavior of reporting a single metric with pre-determined label
+// values.
+type Handle struct {
+	Recorder
+}
+
+// Measurement is used for reporting a batch of metric values.
+type Measurement struct {
+	Handle Handle
+	Value  float64
+}
+
+// Option supports specifying the various metric options.
+type Option func(*Instrument, *[]registry.Option)
 
 // WithDescription applies provided description.
 func WithDescription(desc string) Option {
-	return func(_ *Handle, to *[]registry.Option) {
+	return func(_ *Instrument, to *[]registry.Option) {
```

Comment:
If options are a function that is only called the first time a time series is registered, you can avoid these allocations

Reply:
@krnowak has suggested an options struct. I think I prefer that. Would you object?

Reply:
Options struct sounds like a good idea as well.